File: app/src/main/java/com/fortitude/shamsulkarim/ieltsfordory/ui/MainActivity.java
package com.fortitude.shamsulkarim.ieltsfordory.ui;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.pm.ActivityInfo;
import android.content.res.Configuration;
import android.os.Bundle;
import android.view.Window;
import android.widget.Toast;
import androidx.activity.OnBackPressedCallback;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import com.fortitude.shamsulkarim.ieltsfordory.R;
import com.fortitude.shamsulkarim.ieltsfordory.data_old.FavLearnedState;
import com.fortitude.shamsulkarim.ieltsfordory.domain.auth.usecase.IsUserAuthenticatedUseCase;
import com.fortitude.shamsulkarim.ieltsfordory.domain.auth.usecase.GetCurrentUserUseCase;
import com.fortitude.shamsulkarim.ieltsfordory.data_old.sync.FirebaseSyncManager;
import com.fortitude.shamsulkarim.ieltsfordory.domain.database.usecase.UpdateUserDataUseCase;
import com.fortitude.shamsulkarim.ieltsfordory.domain.database.usecase.AddChildEventListenerUseCase;
import com.fortitude.shamsulkarim.ieltsfordory.domain.learning.usecase.GetFavLearnedStateUseCase;
import org.koin.java.KoinJavaComponent;
import com.fortitude.shamsulkarim.ieltsfordory.ui.words.AllWordsFragment;
import com.fortitude.shamsulkarim.ieltsfordory.ui.favorites.FavoriteFragment;
import com.fortitude.shamsulkarim.ieltsfordory.ui.fragments.HomeFragment;
import com.fortitude.shamsulkarim.ieltsfordory.ui.learned.LearnedFragment;
import com.fortitude.shamsulkarim.ieltsfordory.ui.fragments.ProfileFragment;
import com.fortitude.shamsulkarim.ieltsfordory.domain.connectivity.usecase.IsConnectedUseCase;
import com.google.android.material.bottomnavigation.BottomNavigationView;
import com.google.firebase.auth.FirebaseUser;



public class MainActivity extends AppCompatActivity {

    private BottomNavigationView bottomNavigationView;
    private UpdateUserDataUseCase updateUserDataUseCase;
    private GetFavLearnedStateUseCase getFavLearnedStateUseCase;
    private IsUserAuthenticatedUseCase isUserAuthenticatedUseCase;
    private GetCurrentUserUseCase getCurrentUserUseCase;
    private FirebaseSyncManager syncManager;
    private Toast toast;
    private long lastBackPressTime = 0;
    private boolean connected;
    private IsConnectedUseCase isConnectedUseCase;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        requestWindowFeature(Window.FEATURE_NO_TITLE);
        setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
        setContentView(R.layout.activity_main);

        int screenSize = getResources().getConfiguration().screenLayout &
                Configuration.SCREENLAYOUT_SIZE_MASK;

        String toastMsg;
        switch (screenSize) {
            case Configuration.SCREENLAYOUT_SIZE_LARGE:
                toastMsg = "Large";
                break;
            case Configuration.SCREENLAYOUT_SIZE_NORMAL:
                toastMsg = "Normal";
                break;
            case Configuration.SCREENLAYOUT_SIZE_SMALL:
                toastMsg = "Small screen";
                break;
            case Configuration.SCREENLAYOUT_SIZE_XLARGE:
                toastMsg = "Xlarge";
                break;
            default:
                toastMsg = "Screen size is neither large, normal or small";
        }

        isConnectedUseCase = KoinJavaComponent.get(IsConnectedUseCase.class);
        connected = isConnectedUseCase.execute();

        // BottomNavigation bottomNavigation;
        SharedPreferences sp = this.getSharedPreferences("com.example.shamsulkarim.vocabulary", Context.MODE_PRIVATE);

        if (!sp.contains("soundState")) {

            sp.edit().putBoolean("soundState", true).apply();
            sp.edit().putInt("totalCorrects", 0).apply();
            sp.edit().putInt("noshowads", 0).apply();

        }

        // Initialize repositories
        isUserAuthenticatedUseCase = KoinJavaComponent.get(IsUserAuthenticatedUseCase.class);
        getCurrentUserUseCase = KoinJavaComponent.get(GetCurrentUserUseCase.class);
        updateUserDataUseCase = KoinJavaComponent.get(UpdateUserDataUseCase.class);
        getFavLearnedStateUseCase = KoinJavaComponent.get(GetFavLearnedStateUseCase.class);
        AddChildEventListenerUseCase addChildEventListenerUseCase = KoinJavaComponent.get(AddChildEventListenerUseCase.class);
        syncManager = new FirebaseSyncManager(this, addChildEventListenerUseCase);

        // Initialize bottom navigation
        bottomNavigationView = findViewById(R.id.bottom_navigation);
        setupBottomNavigation();

        // firebase auto sync
        if (isUserAuthenticatedUseCase.execute() && connected) {
            try {
                FirebaseUser currentUser = getCurrentUserUseCase.execute();
                if (currentUser != null) {
                    syncManager.startSync(currentUser.getUid(), null);
                }
            } catch (NullPointerException n) {
                Toast.makeText(this, "Reference exception", Toast.LENGTH_SHORT).show();
            }
        }

        getOnBackPressedDispatcher().addCallback(new OnBackPressedCallback(true) {
            @Override
            public void handleOnBackPressed() {

                if (lastBackPressTime < System.currentTimeMillis() - 4000) {
                    toast = Toast.makeText(getApplicationContext(), "Press back again to close this app",
                            Toast.LENGTH_LONG);
                    toast.show();
                    lastBackPressTime = System.currentTimeMillis();

                } else {
                    if (toast != null) {
                        toast.cancel();
                    }
                    finish();
                }
            }
        });

    }

    // Firebase
    // ----------------------------------------------------------------------------------------------
    private void updateFirebase() {
        SharedPreferences sp = this.getSharedPreferences("com.example.shamsulkarim.vocabulary", Context.MODE_PRIVATE);
        String userName = sp.getString("userName", "Boo");

        // Get aggregated state from LearningProgressRepository
        FavLearnedState favLearnedState = getFavLearnedStateUseCase.execute(userName);

        // Upload to Firebase
        try {
            if (isUserAuthenticatedUseCase.execute()) {
                updateUserDataUseCase.execute(getCurrentUserUseCase.execute().getUid(), favLearnedState, null);
            }
        } catch (Exception e) {
            Toast.makeText(this, "update failure", Toast.LENGTH_SHORT).show();
        }
    }

    // Overriden Methods
    @Override
    protected void onStop() {
        super.onStop();

        if (isUserAuthenticatedUseCase.execute() && connected) {
            updateFirebase();
        }
    }

    private void setupBottomNavigation() {
        bottomNavigationView.setOnItemSelectedListener(item -> {
            Fragment selectedFragment = null;
            int itemId = item.getItemId();

            if (itemId == R.id.navigation_home) {
                selectedFragment = new HomeFragment();
            } else if (itemId == R.id.navigation_words) {
                selectedFragment = new AllWordsFragment();
            } else if (itemId == R.id.navigation_learned) {
                selectedFragment = new LearnedFragment();
            } else if (itemId == R.id.navigation_favorite) {
                selectedFragment = new FavoriteFragment();
            } else if (itemId == R.id.navigation_profile) {
                selectedFragment = new ProfileFragment();
            }

            if (selectedFragment != null) {
                getSupportFragmentManager()
                        .beginTransaction()
                        .replace(R.id.frag, selectedFragment)
                        .commit();
                return true;
            }
            return false;
        });

        // Set default selection (Home)
        bottomNavigationView.setSelectedItemId(R.id.navigation_home);
    }

    private boolean isOnline() {
        return isConnectedUseCase.execute();
    }
}

File: app/src/main/java/com/fortitude/shamsulkarim/ieltsfordory/ui/MainActivity.kt
package com.fortitude.shamsulkarim.ieltsfordory.ui

import android.content.Context
import android.content.pm.ActivityInfo
import android.os.Bundle
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.OnBackPressedCallback
import androidx.activity.compose.setContent
import androidx.activity.enableEdgeToEdge
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.ui.Modifier
import com.fortitude.shamsulkarim.ieltsfordory.data_old.FavLearnedState
import com.fortitude.shamsulkarim.ieltsfordory.data_old.sync.FirebaseSyncManager
import com.fortitude.shamsulkarim.ieltsfordory.domain.auth.usecase.GetCurrentUserUseCase
import com.fortitude.shamsulkarim.ieltsfordory.domain.auth.usecase.IsUserAuthenticatedUseCase
import com.fortitude.shamsulkarim.ieltsfordory.domain.connectivity.usecase.IsConnectedUseCase
import com.fortitude.shamsulkarim.ieltsfordory.domain.database.usecase.AddChildEventListenerUseCase
import com.fortitude.shamsulkarim.ieltsfordory.domain.database.usecase.UpdateUserDataUseCase
import com.fortitude.shamsulkarim.ieltsfordory.domain.learning.usecase.GetFavLearnedStateUseCase
import com.fortitude.shamsulkarim.ieltsfordory.ui.navigation.AppNavigation
import com.fortitude.shamsulkarim.ieltsfordory.ui.theme.VocabularyTheme
import org.koin.android.ext.android.inject
import com.fortitude.shamsulkarim.ieltsfordory.data.preferences.ThemeRepository
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.foundation.isSystemInDarkTheme

/**
 * Main Activity using Jetpack Compose.
 * Hosts the AppNavigation composable with bottom navigation.
 */
class MainActivity : ComponentActivity() {

    // Use cases injected via Koin
    private val isConnectedUseCase: IsConnectedUseCase by inject()
    private val isUserAuthenticatedUseCase: IsUserAuthenticatedUseCase by inject()
    private val getCurrentUserUseCase: GetCurrentUserUseCase by inject()
    private val updateUserDataUseCase: UpdateUserDataUseCase by inject()
    private val getFavLearnedStateUseCase: GetFavLearnedStateUseCase by inject()
    private val addChildEventListenerUseCase: AddChildEventListenerUseCase by inject()
    private val themeRepository: ThemeRepository by inject()

    private lateinit var syncManager: FirebaseSyncManager
    private var toast: Toast? = null
    private var lastBackPressTime: Long = 0
    private var isConnected: Boolean = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        
        // Lock to portrait orientation
        requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
        
        // Enable edge-to-edge display
        enableEdgeToEdge()

        // Check connectivity
        isConnected = isConnectedUseCase.execute()

        // Initialize default SharedPreferences if needed
        initializeDefaultPreferences()

        // Initialize Firebase sync manager
        syncManager = FirebaseSyncManager(this, addChildEventListenerUseCase)

        // Start Firebase auto-sync if authenticated and connected
        startFirebaseSync()

        // Setup double-back-to-exit behavior
        setupBackPressHandler()

        // Set Compose content
        setContent {
            val themeMode by themeRepository.themeMode.collectAsState(initial = 0)
            val isSystemDark = isSystemInDarkTheme()
            val isDarkTheme = when (themeMode) {
                0 -> false // Light
                1 -> true  // Dark
                else -> isSystemDark // System
            }

            VocabularyTheme(darkTheme = isDarkTheme) {
                AppNavigation(
                    modifier = Modifier.fillMaxSize()
                )
            }
        }
    }

    /**
     * Initialize default SharedPreferences values if not already set.
     */
    private fun initializeDefaultPreferences() {
        val sp = getSharedPreferences("com.example.shamsulkarim.vocabulary", Context.MODE_PRIVATE)
        
        if (!sp.contains("soundState")) {
            sp.edit().apply {
                putBoolean("soundState", true)
                putInt("totalCorrects", 0)
                putInt("noshowads", 0)
                apply()
            }
        }
    }

    /**
     * Start Firebase sync if user is authenticated and connected.
     */
    private fun startFirebaseSync() {
        if (isUserAuthenticatedUseCase.execute() && isConnected) {
            try {
                val currentUser = getCurrentUserUseCase.execute()
                currentUser?.let { user ->
                    syncManager.startSync(user.uid, null)
                }
            } catch (e: NullPointerException) {
                Toast.makeText(this, "Reference exception", Toast.LENGTH_SHORT).show()
            }
        }
    }

    /**
     * Upload user data to Firebase.
     */
    private fun updateFirebase() {
        val sp = getSharedPreferences("com.example.shamsulkarim.vocabulary", Context.MODE_PRIVATE)
        val userName = sp.getString("userName", "Boo") ?: "Boo"

        // Get aggregated state from LearningProgressRepository
        val favLearnedState: FavLearnedState = getFavLearnedStateUseCase.execute(userName)

        // Upload to Firebase
        try {
            if (isUserAuthenticatedUseCase.execute()) {
                getCurrentUserUseCase.execute()?.let { user ->
                    updateUserDataUseCase.execute(user.uid, favLearnedState, null)
                }
            }
        } catch (e: Exception) {
            Toast.makeText(this, "Update failure", Toast.LENGTH_SHORT).show()
        }
    }

    /**
     * Setup double-back-to-exit confirmation.
     */
    private fun setupBackPressHandler() {
        onBackPressedDispatcher.addCallback(this, object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                val currentTime = System.currentTimeMillis()
                
                if (lastBackPressTime < currentTime - 4000) {
                    toast = Toast.makeText(
                        applicationContext,
                        "Press back again to close this app",
                        Toast.LENGTH_LONG
                    )
                    toast?.show()
                    lastBackPressTime = currentTime
                } else {
                    toast?.cancel()
                    finish()
                }
            }
        })
    }

    override fun onStop() {
        super.onStop()
        
        // Sync data to Firebase when leaving the app
        if (isUserAuthenticatedUseCase.execute() && isConnected) {
            updateFirebase()
        }
    }
}



